### ISSUE TITLE
Integrity error: Could not sign '/buckets/staging/collections/addons'

### ISSUE THREAD

**Issue by leplatrem** [Kinto/kinto-signer#156]
Looks as if two requests were trying to update `/buckets/monitor/collections/changes` at the same time...

![screenshot from 2017-01-13 16-01-31](https://cloud.githubusercontent.com/assets/546692/21934409/947de592-d9a9-11e6-8190-844a8252e40e.png)

```
IntegrityError: (psycopg2.IntegrityError) duplicate key value violates unique constraint "idx_records_parent_id_collection_id_last_modified"
DETAIL:  Key (parent_id, collection_id, last_modified)=(/buckets/monitor/collections/changes, record, 2017-01-13 14:00:29.794) already exists.
 [SQL: '\n        WITH delete_potential_tombstone AS (\n            DELETE FROM deleted\n             WHERE id = %(object_id)s\n               AND parent_id = %(parent_id)s\n               AND collection_id = %(collection_id)s\n        )\n        INSERT INTO records (id, parent_id, collection_id, data, last_modified)\n        VALUES (%(object_id)s, %(parent_id)s,\n                %(collection_id)s, (%(data)s)::JSONB,\n                from_epoch(%(last_modified)s))\n        RETURNING as_epoch(last_modified) AS last_modified;\n        '] [parameters: {'parent_id': '/buckets/monitor/collections/changes', 'last_modified': 1484316029794L, 'data': '{"host":"firefox.settings.services.mozilla.com","bucket":"blocklists","collection":"addons"}', 'object_id': u'b7f595f9-5fc5-d863-b5dd-e5425dcf427a', 'collection_id': 'record'}]
```

Related to https://github.com/Kinto/kinto/issues/602 ?

**Comment by leplatrem** [comment]
It is related to this: https://github.com/mozilla-services/puppet-config/pull/2430 

When kinto-signer is used together with kinto-changes, we can end-up in a situation where the preview collection monitor/changes record will have the same timestamp as the signed collection monitor/change record. Two distinct records cannot have the same timestamp.

```
SELECT id, data, as_epoch(last_modified) FROM records WHERE parent_id='/buckets/monitor/collections/changes';
                  id                  |                                     data                                     |   as_epoch    
--------------------------------------+------------------------------------------------------------------------------+---------------
 d7ba7803-8088-d1c9-0cce-a68e5510b4e0 | {"host": null, "bucket": "blocklists-preview", "collection": "plugins"}      | 1484328251927
 18aecaeb-52b4-edb9-00c9-d33dfcd337cb | {"host": null, "bucket": "blocklists-preview", "collection": "gfx"}          | 1484328248510
 e46c7c44-36a4-a7e7-2a53-f6829697957e | {"host": null, "bucket": "blocklists", "collection": "gfx"}                  | 1484328214414
 dd40a74e-d3be-276e-e65d-0e76046f8c34 | {"host": null, "bucket": "blocklists", "collection": "plugins"}              | 1484328219406
 1557b6d3-8873-3315-1e80-fd5a1eee394d | {"host": null, "bucket": "blocklists-preview", "collection": "certificates"} | 1484328260537
 4268022b-a9e2-fd58-361d-15cd2fdd0b81 | {"host": null, "bucket": "blocklists", "collection": "certificates"}         | 1484328228265
 7506f4de-77b0-47aa-7812-1752b2445492 | {"host": null, "bucket": "blocklists-preview", "collection": "addons"}       | 1484328243327
 7508dfce-8d5e-b114-45f2-8b4e1880f181 | {"host": null, "bucket": "blocklists", "collection": "addons"}               | 1484328242264

```


```
BackendError: IntegrityError: (psycopg2.IntegrityError) duplicate key value violates unique constraint "idx_records_parent_id_collection_id_last_modified"
DETAIL:  Key (parent_id, collection_id, last_modified)=(/buckets/monitor/collections/changes, record, 2017-01-13 17:24:03.327) already exists.
 [SQL: '\n        UPDATE records SET data=(%(data)s)::JSONB,\n                           last_modified=from_epoch(%(last_modified)s)\n        WHERE id = %(object_id)s\n           AND parent_id = %(parent_id)s\n           AND collection_id = %(collection_id)s\n        RETURNING as_epoch(last_modified) AS last_modified;\n        '] [parameters: {'parent_id': '/buckets/monitor/collections/changes', 'last_modified': 1484328243327L, 'data': '{"host":null,"bucket":"blocklists","collection":"addons"}', 'object_id': u'7508dfce-8d5e-b114-45f2-8b4e1880f181', 'collection_id': 'record'}]
```


I can reproduce (didn't no dig out too much), with the following script on kinto-dist:

```
SERVER="${SERVER:-http://localhost:8888/v1}"
AUTH="${AUTH:-user:pass}"
EDITOR_AUTH="${EDITOR_AUTH:-editor:pass}"
REVIEWER_AUTH="${REVIEWER_AUTH:-reviewer:pass}"

http POST http://localhost:8888/v1/__flush__

echo '{"permissions": {"write": ["system.Authenticated"]}}' | http PUT $SERVER/buckets/staging --auth="$AUTH"
python create_groups.py --bucket=staging --auth="$AUTH" --editor-auth="$EDITOR_AUTH" --reviewer-auth="$REVIEWER_AUTH"
json2kinto --server $SERVER --addons-server https://addons.mozilla.org/ -S amo-blocklist.json --auth="$AUTH" --editor-auth="$EDITOR_AUTH" --reviewer-auth="$REVIEWER_AUTH"

record_id=$(http "$SERVER/buckets/blocklists/collections/addons/records?_limit=1" | python -m json.tool | grep '"id"' | cut -d\" -f4)

http DELETE $SERVER/buckets/staging/collections/addons/records/$record_id --auth="$AUTH"

json2kinto --server $SERVER --addons-server https://addons.allizom.org/ -S amo-blocklist.json --auth="$EDITOR_AUTH"
json2kinto --server $SERVER --addons-server https://addons.allizom.org/ -S amo-blocklist.json --auth="$EDITOR_AUTH"
json2kinto --server $SERVER --addons-server https://addons.allizom.org/ -S amo-blocklist.json --auth="$EDITOR_AUTH"
json2kinto --server $SERVER --addons-server https://addons.allizom.org/ -S amo-blocklist.json --auth="$EDITOR_AUTH"

echo '{"data":{"status":"to-review"}}' | http PATCH :8888/v1/buckets/staging/collections/addons --auth editor:pass
echo '{"data":{"status":"to-sign"}}' | http PATCH :8888/v1/buckets/staging/collections/addons --auth reviewer:pass

```

**Comment by leplatrem** [comment]
Chosen workaround in production: 
* Disable monitoring of changes in preview buckets
* Delete existing monitoring entries for preview buckets with: 

`http DELETE :8888/v1/buckets/monitor/collections/changes/records?bucket=blocklists-preview --auth=user:pass`

(with a user/pass whose principals are listed in `kinto.record_write_principals` setting in INI

**Comment by leplatrem** [comment]
This is probably related to the fact that we keep the `last_modified` timestamps between source, preview and destination. If the collection timestamps ends-up being the same between preview and destination, it fails.

From what I could partially figure out, it happens when some update occurs, a review is requested, and the review approval fails (eg. not in reviewers group). Then when a proper reviewer wants to approve, the copy to the destination is performed, and the collection timestamp becomes the same as the one in preview.

**Comment by leplatrem** [comment]
We fixed the problem in Kinto/kinto-changes#28
